src/features/semanticTokensProvider.ts

```diff
@@ -0,0 +1,315 @@
+    WhiteSpace,
+    Text,
+    StaticSymbol,
+    PreprocessorText,
+    Punctuation,
+    VerbatimStringLiteral,
+    StringEscapeCharacter,
+    ClassName,
+    DelegateName,
+    EnumName,
+    InterfaceName,
+    ModuleName,
+    StructName,
+    TypeParameterName,
+    FieldName,
+    EnumMemberName,
+    ConstantName,
+    LocalName,
+    ParameterName,
+    MethodName,
+    ExtensionMethodName,
+    PropertyName,
+    EventName,
+    NamespaceName,
+    LabelName,
+    XmlDocCommentAttributeName,
+    XmlDocCommentAttributeQuotes,
+    XmlDocCommentAttributeValue,
+    XmlDocCommentCDataSection,
+    XmlDocCommentComment,
+    XmlDocCommentDelimiter,
+    XmlDocCommentEntityReference,
+    XmlDocCommentName,
+    XmlDocCommentProcessingInstruction,
+    XmlDocCommentText,
+    XmlLiteralAttributeName,
+    XmlLiteralAttributeQuotes,
+    XmlLiteralAttributeValue,
+    XmlLiteralCDataSection,
+    XmlLiteralComment,
+    XmlLiteralDelimiter,
+    XmlLiteralEmbeddedExpression,
+    XmlLiteralEntityReference,
+    XmlLiteralName,
+    XmlLiteralProcessingInstruction,
+    XmlLiteralText,
+    RegexComment,
+    RegexCharacterClass,
+    RegexAnchor,
+    RegexQuantifier,
+    RegexGrouping,
+    RegexAlternation,
+    RegexText,
+    RegexSelfEscapedCharacter,
+    RegexOtherEscape,
+}
+
+enum SemanticHighlightModifier {
+    Static
+}
+
+export default class SemanticTokensProvider extends AbstractProvider implements vscode.DocumentSemanticTokensProvider, vscode.DocumentRangeSemanticTokensProvider {
+
+    getLegend(): vscode.SemanticTokensLegend {
+        return new vscode.SemanticTokensLegend(tokenTypes, tokenModifiers);
+    }
+
+    async provideDocumentSemanticTokens(document: vscode.TextDocument, token: vscode.CancellationToken): Promise<vscode.SemanticTokens | null> {
+
+        return this._provideSemanticTokens(document, null, token);
+    }
+
+    async provideDocumentRangeSemanticTokens(document: vscode.TextDocument, range: vscode.Range, token: vscode.CancellationToken): Promise<vscode.SemanticTokens | null> {
+
+        const v2Range: protocol.V2.Range = {
+            Start: {
+                Line: range.start.line,
+                Column: range.start.character
+            },
+            End: {
+                Line: range.end.line,
+                Column: range.end.character
+            }
+        };
+        return this._provideSemanticTokens(document, v2Range, token);
+    }
+
+    async _provideSemanticTokens(document: vscode.TextDocument, range: protocol.V2.Range, token: vscode.CancellationToken): Promise<vscode.SemanticTokens | null> {
+        let req = createRequest<protocol.V2.SemanticHighlightRequest>(document, new vscode.Position(0, 0));
+        req.Range = range;
+
+        let response = await serverUtils.getSemanticHighlights(this._server, req);
+
+        const semanticSpans = response.Spans;
+
+        const builder = new vscode.SemanticTokensBuilder();
+        for (let span of semanticSpans) {
```

Comment:
Maybe I'm misunderstanding the code but why not have the server just return the LSP compatible type vs. re-building it here?

Reply:
I was just following the [typescript extension](https://github.com/microsoft/vscode/blob/master/extensions/typescript-language-features/src/features/semanticTokens.ts#L65) as an example where they request information from their server then perform some post processing. I'll have to think more on whether to keep it like this.

Reply:
Ah, TypeScript hasn't moved fully over to LSP yet which is why they follow that model. They plan to move fully over to LSP